### PR TITLE
Removed SLEAP dependency, updated instructions

### PR DESCRIPTION
**Pull request by sean-afshar**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,13 +32,51 @@ conda env remove -n sleap-anipose
...
+
+1. Activate the conda environment (assuming you installed via conda).
+```
+conda activate sleap-anipose
+```
+
+2. Generate a calibration board matching the details of the board presented before the cameras. 
+
+```
+slap-write_board --fname board.toml --board_width 8 --board_height 11 --square_length 24.0 marker_length --18.5 --marker_bits 4 dict_size 1000
```
Comment (talmo) [inline README.md]: is this right? `marker_length --18.5`

File: README.md [hunk 2]
```diff
@@ -32,13 +32,51 @@ conda env remove -n sleap-anipose
...
+
+4. Calibrate your camera setup by displaying the calibration board to all cameras. Make sure to do this before running any experiments. See [`CALIBRATION_GUIDE.md`](docs/CALIBRATION_GUIDE.md) for more information.
+
+5. Track your data using SLEAP. Make sure that animal tracks are matching across multiple views. See [`SLEAP_GUIDE.md`](docs/SLEAP_GUIDE.md) for more information.
+
+6. Set up your data according to [`FOLDER_STRUCTURE.md`](docs/FOLDER_STRUCTURE.md).
+
+7. Generate the calibration for your cameras:
+```
+slap-calibrate --session date_directory/session_directory --board board.toml --save_calib True --save_folder date_directory/session_directory --save_metadata True --histogram True -reproj_imgs True --save_hist True --save_reproj_imgs True
```
Comment (talmo) [inline README.md:63]: if you have boolean flags, just use the `is_flag=True` option in the decorator so you can make the CLI less verbose ([example](https://click.palletsprojects.com/en/8.1.x/options/#:~:text=For%20single%20option%20boolean%20flags%2C%20the%20default%20remains%20hidden%20if%20the%20default%20value%20is%20False.))

File: README.md [hunk 3]
```diff
@@ -32,13 +32,51 @@ conda env remove -n sleap-anipose
...
 ```python
 import sleap_anipose as slap
 
 session = "path/to/data"
 
-slap.calibrate(session)
-slap.triangulate(session)
+slap.calibrate(session, *calibration_args)
+slap.triangulate(session, *triangulation_args)
 ```
```
Comment (talmo) [inline README.md]: no args unpacking! be explicit with the args. the user has no way of knowing what `*calibration_args` is supposed to be.

File: docs/CALIBRATION_GUIDE.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+The purpose of camera calibration is to gain a quantitative understanding of how the cameras relate to each other in space and how they differ from each other due to their internal mechanisms. 
+To model the spatial relationships of cameras and their internal specifications, we rely on 2 sets of matrices for each camera, the extrinsics and intrinsics matrices respectively. Once we have these sets of matrices, we can then triangulate 3D poses from our tracked 2D poses. 
+
+# Practical Considerations
+
+## Calibration Board 
+
+A calibration board is a rectangular surface with a recognizable pattern that allows for quick detection via modern computer vision software. The most commonly used types of boards are checkerboards, aruco boards (which are rectangular grids with unique markers), and charuco boards (which are a combination of the two). 
+
+**INSERT IMAGE HERE**
```
Comment (talmo) [inline docs/CALIBRATION_GUIDE.md]: Comment this out with a description if you want to have a placeholder, for example:
```
<!-- TODO: Insert screenshot of example board -->
```

File: docs/CALIBRATION_GUIDE.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+1. Through the CLI
+
+```
+slap-write_board --board_name board --img_format jpg --board_X 8 --board_Y 11 --square_length 24.0 --marker_length --18.75
+```
+
+2. Through the API
+
+```
+python
```
Comment (talmo) [inline docs/CALIBRATION_GUIDE.md]: I think you meant for `python` to be on the same line as the backticks to enable syntax highlighting?

File: docs/CALIBRATION_GUIDE.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+2. Through the API
+
+```
+python 
+
+import sleap-anipose as slap 
+slap.calibration.write_board("path/board", 'jpg', 8, 11, 24.0, 18.75)
+```
+
+Currently, we only support charuco board calibration, but plan to expand to checkerboards and aruco boards. It is also important to note that we currently only generate aruco markers with 4 bits from a size 1000 dictionary. However, if one used a calibration board with a different aruco encoding the board can still be used for calibration and triangulation using sleap-anipose. One could do so by writing a toml file that describes the board according to the attributes in the write_board function. Refer to the [`API`](sleap_anipose/calibration.py) for more details.
```
Comment (talmo) [inline docs/CALIBRATION_GUIDE.md]: "API" doesn't need backticks here -- it's not code or a symbol. Also it should say "function documentation", not API.

File: docs/CALIBRATION_GUIDE.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+**INCLUDE IMAGE HERE**
+
+3. Avoid placing the cameras in positions that fail to capture the full 3D features of the animal.
+
+**INCLUDE IMAGE HERE**
+
+4. Make sure that there is enough space alotted in the experimental space for the cameras to capture videos of the calibration board at various angles and at a reasonable resolution. 
+
+**INCLUDE IMAGE HERE**
```
Comment (talmo) [inline docs/CALIBRATION_GUIDE.md]: See comments above about placeholder comments.

File: docs/CALIBRATION_GUIDE.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+Last but not least, it is vital to have the cameras synchronized when recording. This can be achieved with a hardware controller (such as a Rasberry Pi) and a pulse emitter. The synchronized videos must then be saved by an appropriate acquisition software and compressed if necessary. For the benchmark dataset, the camera setup consisted of 5 (4 for some sessions) FLIR Blackfly S Mono cameras, each with ThorLabs fixed focal length 3.5 or 4.5 mm lenses (MVL4WA/ 5WA). Cameras were triggered with a strobe pulse from a tdt rx8. Data was acquired with spinview software and then compressed with ffmpeg.
```
Comment (talmo) [inline docs/CALIBRATION_GUIDE.md]: Awesome!

File: docs/CALIBRATION_GUIDE.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+3. Start recording. 
+
+4. Move the calibration board around the setup to expose it to other cameras from which it may have been occluded. Even if the board is visible to all cameras one should still move the board around so that marker detection can be robust to variability in board position. Also, it is critical to move the board around at a pace that is suitable for the framerate of the cameras, as blurry frames lead to poor detection. 
+
+5. Stop recording. 
+
+6. Check the videos for proper synchronization and visibility, troubleshoot if necessary. 
+
+Alternatively, instead of taking videos for board calibration, one could take synchronized images from each view and then use sleap-anipose to stitch the frames into a movie. Refer to the [`API`](sleap_anipose/calibration.py) for more details.
```
Comment (talmo) [inline docs/CALIBRATION_GUIDE.md]: See above comment about calling this link an "API"

File: docs/FOLDER_STRUCTURE.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
+# Folder structure
+
+This package expects that your data has already been pose tracked and proofread with
+[SLEAP](https://sleap.ai), and organized according to the specifications below.
+
+**It is very important to note that any file ending in an asterisk can be generated from running certain functions in the triangulation and calibration modules. Files and folders not ending in an asterisk must exist before running any functions from sleap-anipose, and must exist in the structure detailed below.**
```
Comment (talmo) [inline docs/FOLDER_STRUCTURE.md:8]: Maybe it's better to have two example folder structures: one with just the initial files at the start of the workflow (before any sleap-anipose steps), and one with everything, including the outputs at the end of the pipeline.

File: sleap_anipose/calibration.py [hunk 11]
```diff
@@ -351,6 +352,75 @@ def write_board_cli(
...
+        marker_length: Length of marker edges in meters.
+        marker_bits: Number of bits in aruco markers, default 4.
+        dict_size: Size of dictionary for encoding aruco markers, default 1000.
+    """
+    # Add more options for aruco dictionaries later.
+    if marker_bits == 4:
+        if dict_size == 1000:
+            aruco_dict = cv2.aruco.Dictionary_get(cv2.aruco.DICT_4X4_1000)
+    else:
+        aruco_dict = cv2.aruco.Dictionary_get(cv2.aruco.DICT_4X4_1000)
```
Comment (talmo) [inline sleap_anipose/calibration.py]: ???

This is the same as the first part of the if statement?

If this is a placeholder, do something like:

```
        if dict_size == 1000:
            aruco_dict = cv2.aruco.Dictionary_get(cv2.aruco.DICT_4X4_1000)
        else:
            # TODO: generate non-standard size
            pass
```

File: sleap_anipose/calibration.py [hunk 12]
```diff
@@ -351,6 +352,75 @@ def write_board_cli(
...
+
+    charuco_board = cv2.aruco.CharucoBoard_create(
+        board_X, board_Y, square_length, marker_length, aruco_dict
+    )
+    imboard = charuco_board.draw((1080, 1080))
```
Comment (talmo) [inline sleap_anipose/calibration.py]: What are these numbers?? should they be function arguments?

File: docs/FOLDER_STRUCTURE.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+- Session directories in the benchmark dataset are named for the date they are recorded on and then the time within the day the recording started.
+
+    For example, session 04182022182439 can be deduced as having started on 04-18-2022 at 18 hours, 24 minutes, 39 seconds. 
+
+- View directories are named are the camera views they represent. 
+
+- The video found in each view folder is named following the convention: `{view}-{session}-{suffix}.mp4`. The '0000_h265_CRF14_denoised' suffix found in the benchmark dataset is a result of the video rendering software.
+
+- The slp files are named following the convention: `{view}-{session}.predictions.{prediction_suffix}.slp`. If the file contains the proofread tracks, the .proofread suffix is added to the file name.
```
Comment (talmo) [inline docs/FOLDER_STRUCTURE.md:64]: do we need to add some mention of the analysis.h5 files?

**Review by talmo — CHANGES_REQUESTED** [review]

File: docs/FOLDER_STRUCTURE.md (hunk 13 above)
Comment (sean-afshar) [reply docs/FOLDER_STRUCTURE.md:64]: yes, i am re-writing the folder structure file, will update in new pull request

File: sleap_anipose/calibration.py (hunk 12 above)
Comment (sean-afshar) [reply sleap_anipose/calibration.py]: these correspond to the number of pixels in the image along each dimension. Do you prefer to have image size as a user parameter?

File: sleap_anipose/calibration.py (hunk 11 above)
Comment (sean-afshar) [reply sleap_anipose/calibration.py]: that works much better, done!

File: sleap_anipose/calibration.py (hunk 12 above)
Comment (talmo) [reply sleap_anipose/calibration.py]: yes -- no [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) in code!!

File: README.md (hunk 1 above)
Comment (sean-afshar) [reply README.md]: should have been --marker_length 18.5, fixed.

File: README.md (hunk 3 above)
Comment (sean-afshar) [reply README.md]: Will expand this section!

File: sleap_anipose/calibration.py (hunk 12 above)
Comment (sean-afshar) [reply sleap_anipose/calibration.py]: will add these as parameters

Comment (sean-afshar) [reply sleap_anipose/calibration.py]: resolved

File: docs/FOLDER_STRUCTURE.md (hunk 10 above)
Comment (sean-afshar) [reply docs/FOLDER_STRUCTURE.md:8]: You are absolutely right. I'll start with just explaining the basics of what a session folder looks like and then expand from there. That's the key to the organization anyways.

Comment (sean-afshar) [reply docs/FOLDER_STRUCTURE.md:8]: Folder structure to be updated in subsequent PR.

File: docs/CALIBRATION_GUIDE.md (hunk 9 above)
Comment (sean-afshar) [reply docs/CALIBRATION_GUIDE.md]: Corrected

File: docs/CALIBRATION_GUIDE.md (hunk 6 above)
Comment (sean-afshar) [reply docs/CALIBRATION_GUIDE.md]: Corrected

File: README.md (hunk 2 above)
Comment (sean-afshar) [reply README.md:63]: Corrected

File: docs/CALIBRATION_GUIDE.md (hunk 5 above)
Comment (sean-afshar) [reply docs/CALIBRATION_GUIDE.md]: resolved

File: README.md (hunk 3 above)
Comment (sean-afshar) [reply README.md]: Will be updated in next PR, left a comment to fix this in the guide.

File: docs/CALIBRATION_GUIDE.md (hunk 4 above)
Comment (sean-afshar) [reply docs/CALIBRATION_GUIDE.md]: corrected

File: docs/CALIBRATION_GUIDE.md (hunk 7 above)
Comment (sean-afshar) [reply docs/CALIBRATION_GUIDE.md]: corrected

File: README.md [hunk 14]
```diff
@@ -32,13 +32,53 @@ conda env remove -n sleap-anipose
...
 
-slap.calibrate(session)
-slap.triangulate(session)
+slap.calibrate(session, *calibration_args) 
+slap.triangulate(session, *triangulation_args)
 ```
 
-See [`FOLDER_STRUCTURE.md`](FOLDER_STRUCTURE.md) for details on how session data should be organized.
\ No newline at end of file
+# TODO: unpack function args and explain them.
```
Comment (talmo) [inline README.md]: In markdown, write comments like `<!-- TODO: unpack function args and explain them.  -->`

File: docs/CALIBRATION_GUIDE.md [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+6. Check the videos for proper synchronization and visibility, troubleshoot if necessary. 
+
+Alternatively, instead of taking videos for board calibration, one could take synchronized images from each view and then use sleap-anipose to stitch the frames into a movie. Refer to the [function documentation](sleap_anipose/calibration.py) for more details. 
+
+7. Organize the different videos into their proper view subfolders as explained in [FOLDER_STRUCTURE.md](sleap-anipose/docs/FOLDER_STRUCTURE.md).
+
+8. Run the calibration function from either the command line or a script.
+
+```
+slap-calibrate -- session my/path --board my/path/board.toml --calib_fname my/path/calibration.toml --metadata_fname my/path/calibration.metadata.h5 --histogram_path my/path --reproj_path my/path
```
Comment (talmo) [inline docs/CALIBRATION_GUIDE.md]: remove extra space after first `--`

File: docs/CALIBRATION_GUIDE.md [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+```
+
+```python
+
+import sleap-anipose as slap 
+cgroup, metadata = slap.calibration.calibrate("my/path", "my/path/board.toml", "my/path/calibration.toml", "my/path/calibration.metadata.h5", "my/path", "my/path")
```
Comment (talmo) [inline docs/CALIBRATION_GUIDE.md]: Fix import and call:
```
import sleap_anipose as slap 
cgroup, metadata = slap.calibrate("my/path", "my/path/board.toml", "my/path/calibration.toml", "my/path/calibration.metadata.h5", "my/path", "my/path")
```

If `calibrate` isn't at the top-level, import it in `__init__.py`.

File: docs/FOLDER_STRUCTURE.md [hunk 17]
```diff
@@ -0,0 +1,71 @@
+# Folder structure
+
+# TODO: Update basic structure tree, build understanding from ground up
```
Comment (talmo) [inline docs/FOLDER_STRUCTURE.md]: Fix markdown comment syntax

File: sleap_anipose/calibration.py [hunk 18]
```diff
@@ -61,13 +62,10 @@ def make_reproj_imgs(
...
-        n_samples: The number of images to make per view.
         session: Path containing the view subfolders with the calibration board
-            images. Images are saved to the view subfolders in this folder as
-            'session / view / reprojection-{frame}.jpg'. Assumed to be the
-            current working directory if not specified.
-        save: Flag determining whether or not to save images to the session.
-
+            images.
+        n_samples: The number of images to make per view.
+        save_path: The session to save the images to. If not specified as a non-empty string, images will nto be saved. Images are saved to the view subfolders in this folder as 'save_path / view / reprojection-{frame}.png'.
```
Comment (talmo) [inline sleap_anipose/calibration.py]: Fix typo ("nto" -> "not") and make sure line length doesn't exceed 88 chars

File: sleap_anipose/calibration.py [hunk 19]
```diff
@@ -124,11 +119,7 @@ def get_metadata(
                 'rvec': A rotation vector for the orientation of the image
                 'tvec': A translation vector for the orientation of the image
         cgroup: The object containing all the relevant camera parameters.
-        save: Flag determining whether to save the calibration metadata
-            to the session. Files will be saved as
-            'session / calibration_metadata.h5'.
-        session: Path to save the metadata to. Assumed to be the current
-            working directory if not specified.
+        save_path: The file path to save the metadata to. Will only save if a non-empty string is entered.
```
Comment (talmo) [inline sleap_anipose/calibration.py]: line length?

File: sleap_anipose/triangulation.py [hunk 20]
```diff
@@ -55,17 +55,22 @@ def triangulate(
...
+                constraints: A Kx2 array array for rigid limb constraints, default empty. An example would be [[0, 1], [2,3]], which denotes that the length between joints 1 and 2 and the length between joints 2 and 3 are constant.
+                constraints_weak: A Kx2 array of more flexible constraints such as shoulder length in humans or tarsus length in flies, default empty.
+                scale_smooth: The weight of the temporal smoothing term in the loss function, default 4.
+                scale_length: The weight of the length constraints in the loss function, default 2.
+                scale_length_weak: The weight of the weak length constraints in the loss function, default 0.5.
+                reproj_error_threshold: A threshold for determining which points are not suitable for triangulation, default 15.
+                reproj_loss: The loss function for the reprojection loss, default 'soft_l1'. See scipy.optimize.least_squares for more options.
+                n_deriv_smooth: The order of derivative to smooth for in the temporal filtering, default 1.
```
Comment (talmo) [inline sleap_anipose/triangulation.py]: Line length

File: sleap_anipose/triangulation.py [hunk 21]
```diff
@@ -119,29 +115,86 @@ def triangulate(
...
 )
 @click.option(
     "--disp_progress",
     default=False,
     help="Flag determining whether or not to display triangulation progress.",
 )
+@click.option(
+    "--constraints",
+    default=[],
+    help="A Kx2 array array for rigid limb constraints, default empty. An example would be [[0, 1], [2,3]], which denotes that the length between joints 1 and 2 and the length between joints 2 and 3 are constant.",
```
Comment (talmo) [inline sleap_anipose/triangulation.py]: Line length

**Review by talmo — CHANGES_REQUESTED** [review]
Minor fixes

File: README.md (hunk 14 above)
Comment (sean-afshar) [reply README.md]: resolved

File: docs/CALIBRATION_GUIDE.md (hunk 15 above)
Comment (sean-afshar) [reply docs/CALIBRATION_GUIDE.md]: good catch!

File: docs/FOLDER_STRUCTURE.md (hunk 17 above)
Comment (sean-afshar) [reply docs/FOLDER_STRUCTURE.md]: resolved

File: sleap_anipose/triangulation.py (hunk 21 above)
Comment (sean-afshar) [reply sleap_anipose/triangulation.py]: resolved

File: sleap_anipose/triangulation.py (hunk 20 above)
Comment (sean-afshar) [reply sleap_anipose/triangulation.py]: resolved

File: sleap_anipose/calibration.py (hunk 19 above)
Comment (sean-afshar) [reply sleap_anipose/calibration.py]: resolved

File: sleap_anipose/calibration.py (hunk 18 above)
Comment (sean-afshar) [reply sleap_anipose/calibration.py]: resolved

File: docs/CALIBRATION_GUIDE.md (hunk 16 above)
Comment (sean-afshar) [reply docs/CALIBRATION_GUIDE.md]: resolved

**Review by talmo — APPROVED** [review]

**Pull request merged**